Add enum types valuemust be the same across source schemas (#42)
graphql/composite-schemas-spec: spec/Section 4 -- Composition.md

### Pre Merge Validation

#### Enum Type Values Must Be The Same Across Source Schemas

**Error Code**

ENUM_VALUES_MUST_BE_THE_SAME_ACROSS_SCHEMAS

**Formal Specification**

- Let {enumsByName} be a map where the key is the name of an enum type, and the value is a list of all enum types from different source schemas with that name.

> - For each {listOfEnum} in {enumsByName}:
>   - {EnumAreMergeable(listOfEnum)} must be true.
> 
> EnumAreMergeable(enums):

Reviewer: ```suggestion
- For each {listOfEnums} in {enumsByName}:
  - {EnumsAreMergeable(listOfEnums)} must be true.

EnumsAreMergeable(enums):
```

- Let {values} be the set of all values of the first enum in {enums}
- For each {enum} in {enums}
  - Let {enumValues} be the set of all values of {enum}
  - {values} must be equal to {enumValues}

**Explanatory Text**

This rule ensures that enum types with the same name across different source schemas in a composite schema have identical sets of values. 

> Enums, must be consistent across source schemas to avoid conflicts and ambiguities in the composite schema.

Author: At the moment enum value have to be strictly equal. We do not differentiate between output, input or combined enums

Reviewer 2: > At the moment enum value have to be strictly equal

Do we plan to explore relaxing this to allow evolution of enums?

Author: Yes we just talked about this in yesterdays wg, we are probably going to use the same pattern that federation uses currently.
union in ouput, intersection in input and exact match if used as in- and output

> Enums, must be consistent across source schemas to avoid conflicts and ambiguities in the composite schema.

Reviewer: ```suggestion
Enums must be consistent across source schemas to avoid conflicts and ambiguities in the composite schema.
```

When an enum is defined with differing values, it can lead to confusion and errors in query execution. 
For instance, a value valid in one schema might be passed to another where it's unrecognized, leading to unexpected behavior or failures. 
This rule prevents such inconsistencies by enforcing that all instances of the same named enum across schemas have an exact match in their values.

> In this example, both subgraphs define `Enum1` with the same value `BAR`, satisfying the rule:

Reviewer: ```suggestion
In this example, both source schemas define `Enum1` with the same value `BAR`, satisfying the rule:
```

**Formal Specification**

- Let {enumNames} be the set of all enum type names across all source schemas.
- For each {enumName} in {enumNames}:
  - Let {enums} be the list of all enum types from different source schemas with the name {enumName}.
  - {EnumsAreMergeable(enums)} must be true.

EnumsAreMergeable(enums):

- Let {inaccessibleValues} be the set of values that are declared as `@inaccessible` in {enums}.

> - Let {requriedValues} be the set of values in {enums} that are not in {inaccessibleValues}.

Reviewer: ```suggestion
- Let {requiredValues} be the set of values in {enums} that are not in {inaccessibleValues}.
```

- For each {enum} in {enums}
  - Let {enumValues} be the set of all values of {enum} that are not in {inaccessibleValues}.

>   - {requriedValues} must be equal to {enumValues}

Reviewer: ```suggestion
  - {requiredValues} must be equal to {enumValues}
```
